src/na/na_ofi.c

```diff
@@ -0,0 +1,2255 @@
+    na_ofi_msg_send_expected,               /* msg_send_expected */
+    na_ofi_msg_recv_expected,               /* msg_recv_expected */
+    na_ofi_mem_handle_create,               /* mem_handle_create */
+    NULL,                                   /* mem_handle_create_segment */
+    na_ofi_mem_handle_free,                 /* mem_handle_free */
+    na_ofi_mem_register,                    /* mem_register */
+    na_ofi_mem_deregister,                  /* mem_deregister */
+    NULL,                                   /* mem_publish */
+    NULL,                                   /* mem_unpublish */
+    na_ofi_mem_handle_get_serialize_size,   /* mem_handle_get_serialize_size */
+    na_ofi_mem_handle_serialize,            /* mem_handle_serialize */
+    na_ofi_mem_handle_deserialize,          /* mem_handle_deserialize */
+    na_ofi_put,                             /* put */
+    na_ofi_get,                             /* get */
+    NULL,                                   /* get_poll_fd */
+    na_ofi_progress,                        /* progress */
+    na_ofi_cancel                           /* cancel */
+};
+
+/*****************/
+/* Local Helpers */
+/*****************/
+
+static int
+na_ofi_getinfo()
+{
+    struct fi_info *hints;
+    struct fi_info *providers = NULL;
+    na_return_t ret = NA_SUCCESS;
+    int rc;
+
+    /* Only the first initiator queries OFI providers */
+    if (HG_UTIL_TRUE != hg_atomic_cas32(&nofi_gdata.nog_init_flag, 0, 1)) {
+        while(nofi_gdata.nog_providers == NULL)
+            usleep(1);
+        goto out;
+    }
+
+    ret = na_ofi_config_init();
+    if (ret != NA_SUCCESS) {
+        NA_LOG_ERROR("na_ofi_config_init failed, ret %d.", ret);
+        goto out;
+    }
+
+    /**
+     * Hints to query && filter providers.
+     *
+     * mode: operational mode, NA_OFI passes in context for communication calls.
+     * ep_type: reliable datagram (connection-less).
+     * caps: capabilities required.
+     * msg_order: guarantee that messages with same tag are ordered.
+     * (FI_ORDER_SAS - Send after send. If set, message send operations,
+     *  including tagged sends, are transmitted in the order submitted relative
+     *  to other message send. If not set, message sends may be transmitted out
+     *  of order from their submission).
+     */
+    hints = fi_allocinfo();
+    if (!hints) {
+        NA_LOG_ERROR("fi_allocinfo failed.\n");
+        ret = NA_NOMEM_ERROR;
+        goto out;
+    }
+
+    hints->mode               = FI_CONTEXT;
+    hints->ep_attr->type      = FI_EP_RDM;
+    /* notes: verbs provider does not support FI_SOURCE and FI_MR_SCALABLE */
+    hints->caps               = FI_TAGGED | FI_RMA | FI_SOURCE;
+    hints->tx_attr->msg_order = FI_ORDER_SAS;
+    hints->rx_attr->msg_order = FI_ORDER_SAS;
+
+    hints->domain_attr->threading        = FI_THREAD_UNSPEC;
+    hints->domain_attr->control_progress = FI_PROGRESS_MANUAL;
```

Comment:
Is this required? We don't have an event queue, so the control operations are synchronous. Therefore, the requirement to poll while av_inserts are pending doesn't arise.